src/app/signin/sections/SignInStep4.tsx

```diff
@@ -1,7 +1,153 @@
-import React from "react";
+"use client";
 
-const SignInStep4 = () => {
-  return <div className="bg-red-300 h-full">SignInStep4</div>;
+import React, { useEffect } from "react";
+import { UseFormGetValues, UseFormSetValue } from "react-hook-form";
+
+import { Button } from "@/components";
+import { Check } from "@/components/icons";
+import useTheme from "@/lib/hooks/useTheme";
+
+import { SignInFormProps } from "../SignIn.controller";
+import useSignInModel from "../SignIn.model";
+import * as GS from "@/styles/GlobalStyle";
+import * as S from "../SignIn.styles";
+
+interface SignInStep4Props {
+  getValues: UseFormGetValues<SignInFormProps>;
+  setValue: UseFormSetValue<SignInFormProps>;
+}
+
+const POLICY_CONTENT_LIST: Array<{
+  content: string;
+  // policyView: WINDOW_VIEWS;
+  required: boolean;
+  formKey: "policy1" | "policy2" | "policy3";
+}> = [
+  {
+    content: "[필수] 개인정보 수집 및 이용 동의",
+    required: true,
+    formKey: "policy1",
+  },
+  { content: "[필수] 위치정보 수집 및 이용 동의", required: true, formKey: "policy2" },
+  {
+    content: "[필수] 만 14세 미만 가입 제한",
+    required: true,
+    formKey: "policy3",
+  },
+];
+
+const SignInStep4 = ({ getValues, setValue }: SignInStep4Props) => {
+  const theme = useTheme();
+
+  const { checkAllPolicy, setCheckAllPolicy, policyCheck, setPolicyCheck } = useSignInModel();
+
+  useEffect(() => {
+    const initState = POLICY_CONTENT_LIST.map(({ formKey }) => getValues(formKey));
+    setPolicyCheck(initState);
+  }, []);
+
+  useEffect(() => {
+    if (policyCheck.every((isTrue) => isTrue)) {
+      setCheckAllPolicy(true);
+    } else {
+      setCheckAllPolicy(false);
+    }
+  }, [policyCheck]);
+
+  const togglePolicyAccept = (index: number) => {
+    const newState = policyCheck.map((c, i) => {
+      if (i === index) {
+        // Increment the clicked counter
+        return !c;
+      } else {
+        // The rest haven't changed
+        return c;
+      }
+    });
+    setPolicyCheck(newState);
+    const prevValue = getValues(POLICY_CONTENT_LIST[index].formKey);
+    setValue(POLICY_CONTENT_LIST[index].formKey, !prevValue);
+  };
+
+  const toggleAllPolicy = () => {
+    if (checkAllPolicy) {
+      const newState = new Array(POLICY_CONTENT_LIST.length).fill(false);
+      setPolicyCheck(newState);
+      POLICY_CONTENT_LIST.forEach(({ formKey }) => setValue(formKey, false));
+    } else {
+      const newState = new Array(POLICY_CONTENT_LIST.length).fill(true);
+      setPolicyCheck(newState);
+      POLICY_CONTENT_LIST.forEach(({ formKey }) => setValue(formKey, true));
```

Comment:
[P4] 개인적으로 요 부분도!.... 네... 🫠 요것도 아주 작은 의견이었습니다

```suggestion
    if (checkAllPolicy) {
      const newState = new Array(POLICY_CONTENT_LIST.length).fill(false);
      setPolicyCheck(newState);
      POLICY_CONTENT_LIST.forEach(({ formKey }) => setValue(formKey, false));
      return;
    }
    
    const newState = new Array(POLICY_CONTENT_LIST.length).fill(true);
    setPolicyCheck(newState);
    POLICY_CONTENT_LIST.forEach(({ formKey }) => setValue(formKey, true));
```

Reply:
요부분은 얼리리턴으로 기저조건을 날리는 것보다는
두 조건 중 하나를 실행한다는 느낌이 커서 if else 를 사용했네요!